packages/appium/lib/extension/manifest.js

```diff
@@ -0,0 +1,570 @@
+   * @type {Promise<void>|undefined}
+   */
+  _reading;
+
+  /**
+   * Sets internal data to a fresh clone of {@link INITIAL_MANIFEST_DATA}
+   *
+   * Use {@link Manifest.getInstance} instead.
+   * @param {string} appiumHome
+   * @private
+   */
+  constructor (appiumHome) {
+    this._appiumHome = appiumHome;
+    this._data = _.cloneDeep(INITIAL_MANIFEST_DATA);
+  }
+
+  /**
+   * Returns a new or existing {@link Manifest} instance, based on the value of `appiumHome`.
+   *
+   * Maintains one instance per value of `appiumHome`.
+   * @param {string} appiumHome - Path to `APPIUM_HOME`
+   * @returns {Manifest}
+   */
+  static getInstance = _.memoize(function _getInstance (
+    appiumHome,
+  ) {
+    return new Manifest(appiumHome);
+  });
+
+  /**
+   * Searches `APPIUM_HOME` for installed extensions and adds them to the manifest.
+   * @param {SyncWithInstalledExtensionsOpts} opts
+   * @returns {Promise<boolean>} `true` if any extensions were added, `false` otherwise.
+   */
+  async syncWithInstalledExtensions ({depthLimit = DEFAULT_SEARCH_DEPTH} = {}) {
+    const walkOpts = _.defaults({depthLimit}, DEFAULT_FIND_EXTENSIONS_OPTS);
+    // this could be parallelized, but we can't use fs.walk as an async iterator
+    let didChange = false;
+    for await (const {stats, path: filepath} of fs.walk(
+      this._appiumHome,
+      walkOpts,
+    )) {
+      if (filepath !== this._appiumHome && stats.isDirectory()) {
+        try {
+          const pkg = await env.readPackageInDir(filepath);
+          if (pkg && isExtension(pkg)) {
+            this.addExtensionFromPackage(
+              pkg,
+              path.join(filepath, 'package.json'),
+            );
+            didChange = true;
+          }
+        } catch {}
+      }
+    }
+    return didChange;
+  }
+
+  /**
+   * Given a path to a `package.json`, add it as either a driver or plugin to the manifest.
+   *
+   * Will _not_ overwrite existing entries.
+   * @template {ExtensionType} ExtType
+   * @param {ExtensionPackageJson<ExtType>} pkgJson
+   * @param {string} pkgPath
+   * @returns {void}
+   */
+  addExtensionFromPackage (pkgJson, pkgPath) {
+    if (!isExtension(pkgJson)) {
+      return;
+    }
+
+    /**
+     * @type {InternalData}
+     */
+    const internal = {
+      pkgName: pkgJson.name,
+      version: pkgJson.version,
+      installType: INSTALL_TYPE_NPM,
+      installSpec: `${pkgJson.name}@${pkgJson.version}`,
+    };
+
+    if (isDriver(pkgJson)) {
+      if (!this._data.drivers[pkgJson.appium.driverName]) {
+        this._data.drivers[pkgJson.appium.driverName] = {
+          ..._.omit(pkgJson.appium, 'driverName'),
+          ...internal,
+        };
+      }
+    } else if (isPlugin(pkgJson)) {
+      if (!this._data.plugins[pkgJson.appium.pluginName]) {
+        this._data.plugins[pkgJson.appium.pluginName] = {
+          ..._.omit(pkgJson.appium, 'pluginName'),
+          ...internal,
+        };
+      }
+    } else {
+      throw new TypeError(
```

Comment:
This function (`addExtensionFromPackage`) is called by `syncWithInstalledExtensions`, which is called by `read()` under certain conditions.  `read()` is called by `init()`, which is called by `main()`, and this `TypeError` is going to bubble up to `main`'s rejection handler.

I'd say the people likely to encounter this error are extension developers. They should know if their extension is invalid. For that reason, I probably wouldn't want to just warn; it'd be too easy to miss the fact that their extension isn't loaded, and this would be a confusing debugging experience.

Reply:
👍